typing/ctype.ml

```diff
@@ -4914,9 +4952,30 @@ let immediacy env typ =
             | _ -> false)
           row.row_fields
       then
-        Type_immediacy.Unknown
+        Type_immediacy.coerce Unknown ~as_:imm
       else
-        Type_immediacy.Always
-  | _ -> Type_immediacy.Unknown
-
-let maybe_pointer_type env typ = not (is_immediate (immediacy env typ))
+        Type_immediacy.coerce Always ~as_:imm
+  | _ -> Type_immediacy.coerce Unknown ~as_:imm
```

Comment:
Note: another way to write this function would be to have a `get_type_immediacy` function that returns `Imm of immediacy | Of_type of type_expr`, and `check_type_immediacy` that depends on `~expected` and loops on the result of `get_type_immediacy`. Personally I would find it clearer, and it would also reduce the number of `.coerce` calls to just one.

Reply:
(I'm not sure why you prefer `check_*_immediate` to `check_*_immediacy`; I guess that one reads like "check the immediacy of foo" and the other "check that foo is immediate", but it's not clear that "check that foo is immediate Unknown" is very intuitive, and the name variation may confuse some people.)

Reply:
I don't see how the get_type_immediacy approach would work in the abstract case. It both checks against the declared immediacy (for speed, and in case expansion fails due to missing CMIs), and checks against an expanded type expression. I also find the interface less clear than "check that this type is immediate".

Regarding "immediate" vs. "immediacy": to me, the switch from noun to verb lined up with the switch from inference to checking ("compute the immediacy of this type" became "check that this type is immediate"). I'm not terribly attached to this change, though.

Reply:
I think it would be good to change the code to something like:
```ocaml
let variant_all_constant row =
  List.for_all
    (function
     | _, (Rpresent (Some _) | Reither (false, _, _, _)) -> false
     | _ -> true)
   row.row_fields

let variant_immediacty row =
  let row = Btype.row_repr row in
  (* if all labels are devoid of arguments, not a pointer *)
  if row.row_closed && variant_all_constant row then Always
  else Unknown
        
let head_immediacy env ty =
  match (repr ty).desc with
  | Tconstr(p, _, _) -> begin
      match Env.find_type p env with
      | exception Not_found -> Unknown
      | decl -> kind_immediacy decl.type_kind
    end
  | Tvariant row -> variant_immediacy row
  | _ -> Unknown
  
let rec check_type_immediate env ty imm =
  let imm_head = head_immediacy env ty in
  match Type_immediacy.coerce imm_head ~as_:imm with
  | Ok () -> Ok ()
  | Error _ as err ->
      match try_expand_head try_expand_once_opt env ty with
      | ty' -> check_type_immediate env ty' imm
      | exception Cannot_expand -> err
```
with the above suggested change to `try_expand_once_opt` handling the unboxed stuff.